File: macos-app/Lexume/Services/ExtractionServiceFactory.swift
import Foundation

/// Picks the extraction backend automatically: Gemini when a key is
/// configured, otherwise the on-device OCR engine chosen in Settings.
enum ExtractionServiceFactory {
    static func make(secrets: SecretsStore = KeychainStore()) -> ExtractionService {
        if let key = secrets.get(.geminiAPIKey), !key.isEmpty {
            return GeminiClient(secrets: secrets)
        }
        let engineRaw = UserDefaults.standard.string(forKey: AppSettings.ocrEngineKey) ?? AppSettings.defaultOCREngine
        let engine = OCREngine(rawValue: engineRaw) ?? .vision
        return LocalExtractionService(engine: engine)
    }
}

File: android-app/app/src/main/java/com/aboooooo57/lexume/network/ExtractionServiceFactory.kt
package com.aboooooo57.lexume.network

import com.aboooooo57.lexume.data.local.SecretKey
import com.aboooooo57.lexume.data.local.SecureKeyStore
import com.aboooooo57.lexume.ocr.MlKitOcrService

/**
 * Picks the extraction backend automatically: Gemini when a key is
 * configured, otherwise on-device ML Kit OCR. Mirrors
 * `Services/ExtractionServiceFactory.swift`. Resolved fresh on every call
 * (not cached) since the key can change at any time via Settings.
 */
class ExtractionServiceFactory(
    private val secureKeyStore: SecureKeyStore,
    private val ocr: MlKitOcrService
) {
    suspend fun make(): ExtractionService {
        val key = secureKeyStore.get(SecretKey.GEMINI_API_KEY)
        return if (!key.isNullOrEmpty()) {
            GeminiClient(secureKeyStore)
        } else {
            LocalExtractionService(ocr)
        }
    }
}
